src/v/model/record.h

```diff
@@ -550,8 +550,11 @@ class record_batch {
     bool compressed() const { return _compressed; }
 
     record_batch copy() const {
-        // copy sets shard id
-        return record_batch(_header.copy(), _records.copy(), _compressed);
+        auto b = record_batch(_header.copy(), _records.copy(), _compressed);
+        if (_decompressed) {
+            b.decompressed(_decompressed->copy());
```

Comment:
```
vassert(!_decompressed->compressed(), "");
b.decompressed(_decompressed->copy());
```

would that make sense to add as an assertion?

Reply:
good idea, done!